File: sys/teton/framebuffer/framebuffer.cpp
#include "framebuffer/framebuffer.h"

Framebuffer::Framebuffer(const yellowstone::FramebufferInfo& info)
    : fb_info_(info), cursor_pos_(0) {
  uint64_t buff_size_bytes = fb_info_.height() * fb_info_.pitch();
  fb_memory_ = mmth::OwnedMemoryRegion::DirectPhysical(fb_info_.address_phys(),
                                                       buff_size_bytes);
  fb_ = reinterpret_cast<uint32_t*>(fb_memory_.vaddr());
}

void Framebuffer::DrawPixel(uint32_t row, uint32_t col, uint32_t pixel) {
  // Div by 4 because pitch is in bytes and fb_ is a 32bit array.
  fb_[(row * fb_info_.pitch() / 4) + col] = pixel;
}

void Framebuffer::DrawGlyph(uint8_t* glyph) {
  uint32_t gl_width = 8;
  uint32_t gl_height = 16;

  for (uint8_t r = 0; r < gl_height; r++) {
    for (uint8_t c = 0; c < gl_width; c++) {
      if (((glyph[r] >> c) % 2) == 1) {
        DrawPixel(r, gl_width - c - 1, 0xFFFF'FFFF);
      } else {
        DrawPixel(r, gl_width - c - 1, 0);
      }
    }
  }
};

File: rust/sys/teton/src/framebuffer.rs
use mammoth::{mem::MemoryRegion, zion::ZError};
use yellowstone::FramebufferInfo;

pub struct Framebuffer {
    fb_info: FramebufferInfo,
    memory_region: MemoryRegion,
}

impl Framebuffer {
    pub fn from_info(fb_info: FramebufferInfo) -> Result<Self, ZError> {
        let size = fb_info.height * fb_info.pitch;
        let memory_region = MemoryRegion::direct_physical(fb_info.address_phys, size)?;
        Ok(Self {
            fb_info,
            memory_region,
        })
    }

    fn draw_pixel(&self, row: u32, col: u32, pixel: u32) {
        let index = row * (self.fb_info.pitch as u32 / 4) + col;
        self.memory_region.mut_slice()[index as usize] = pixel;
    }

    pub fn draw_glyph(&self, glyph: &[u8], row: u32, col: u32) {
        let gl_width = 8;
        let gl_height = 16;

        for r in 0..gl_height {
            for c in 0..gl_width {
                if ((glyph[r] >> c) % 2) == 1 {
                    self.draw_pixel(row + (r as u32), col + (gl_width - c - 1), 0xFFFFFFFF);
                } else {
                    self.draw_pixel(row + (r as u32), col + (gl_width - c - 1), 0);
                }
            }
        }
    }

    pub fn width(&self) -> u32 {
        self.fb_info.width as u32
    }

    pub fn height(&self) -> u32 {
        self.fb_info.height as u32
    }
}
